### PR TITLE
auth_server: Exchange auth code for identity token

### PR DESCRIPTION
**Pull request by conradgrobler**
# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [x] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: examples/authentication/server/src/http_client.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+/// Executes a request and ensures that the response code indicates success.
+async fn execute_request(request: Request<Body>) -> Result<String, Box<dyn error::Error>> {
+    let https = HttpsConnector::new();
+    let client = Client::builder().build::<_, hyper::Body>(https);
+    let response = client.request(request).await?;
+    let status = response.status();
+    info!("Response status: {}", &status);
+    let buffer = aggregate(response).await?;
+    let mut writer: Vec<u8> = vec![];
+    copy(&mut buffer.reader(), &mut writer)?;
```
Comment (tiziano88) [inline examples/authentication/server/src/http_client.rs:61]: Alternatively I think `buffer.to_bytes().to_vec()` should work?

File: examples/authentication/server/src/http_client.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    let result = String::from_utf8(writer)?;
+    info!("Received response: {}", &result);
+    if status.is_success() {
+        Ok(result)
+    } else {
+        Err(Box::new(HttpError::new(status)))
+    }
```
Comment (tiziano88) [inline examples/authentication/server/src/http_client.rs:68]: Move the status check up and with an early return?

File: examples/authentication/server/src/auth_handler.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    authentication_server::{Authentication, AuthenticationServer},
+    AuthParameters, IdentityTokenRequest, IdentityTokenResponse,
+};
+use tonic::{Request, Response, Status};
+
+pub mod proto {
+    tonic::include_proto!("oak.examples.authentication");
+}
+
+use crate::token_exchanger;
```
Comment (tiziano88) [inline examples/authentication/server/src/auth_handler.rs:29]: Move up with the other `use`s.

File: examples/authentication/server/src/http_client.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+};
+use hyper_tls::HttpsConnector;
+use log::info;
+use std::{error, fmt, io::copy};
+
+/// Performs an HTTP GET operation to the specified URI.
+// TODO(#924): Support caching based on the max-age value of the cache-control response header
+pub async fn get(uri: Uri) -> Result<String, Box<dyn error::Error>> {
+    info!("Geting {}", &uri);
+    execute_request(Request::get(uri).body(Body::from(""))?).await
```
Comment (tiziano88) [inline examples/authentication/server/src/http_client.rs:31]: Could you add a comment to explain why the body is empty but cannot be omitted?

File: examples/authentication/server/src/http_client.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    Client, StatusCode,
+};
+use hyper_tls::HttpsConnector;
+use log::info;
+use std::{error, fmt, io::copy};
+
+/// Performs an HTTP GET operation to the specified URI.
+// TODO(#924): Support caching based on the max-age value of the cache-control response header
+pub async fn get(uri: Uri) -> Result<String, Box<dyn error::Error>> {
+    info!("Geting {}", &uri);
```
Comment (tiziano88) [inline examples/authentication/server/src/http_client.rs:30]:
```suggestion
    info!("Getting {}", &uri);
```

File: examples/authentication/server/src/main.rs [hunk 6]
```diff
@@ -90,15 +69,14 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let cert = tokio::fs::read(opt.cert).await?;
     let key = tokio::fs::read(opt.private_key).await?;
     let identity = Identity::from_pem(cert, key);
-    let handler = AuthenticationHandler {
-        client_id: opt.client_id.clone(),
-    };
-
     info!("Starting the auth server at {:?}", address);
```
Comment (tiziano88) [inline examples/authentication/server/src/main.rs:73]: Perhaps check that client_id and client_secret are non-empty and output an appropriate error message

File: examples/authentication/server/src/token_exchanger.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Header, Validation};
+use log::info;
+use serde::{Deserialize, Serialize};
+use std::{error, fmt};
+use url::form_urlencoded;
+
+use crate::http_client;
+
+/// Claims encoded into the identity token
+#[derive(Debug, Serialize, Deserialize)]
```
Comment (tiziano88) [inline examples/authentication/server/src/token_exchanger.rs:32]: Add link https://developers.google.com/identity/protocols/oauth2/openid-connect#an-id-tokens-payload

also perhaps add some comments to the fields, or use friendlier name and override the name of the serialized field using some serde annotation.

Comment (tiziano88) [inline examples/authentication/server/src/token_exchanger.rs:32]: Why do we need both Ser and De?

File: examples/authentication/server/src/token_exchanger.rs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+#[derive(Deserialize, Debug)]
+struct TokenResponse {
+    expires_in: usize,
+    id_token: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct KeyList {
```
Comment (tiziano88) [inline examples/authentication/server/src/token_exchanger.rs:50]: Document.

File: examples/authentication/server/src/token_exchanger.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    keys: Vec<Key>,
+}
+
+#[derive(Clone, Deserialize, Debug)]
+struct Key {
```
Comment (tiziano88) [inline examples/authentication/server/src/token_exchanger.rs:55]: Add link to https://auth0.com/docs/tokens/references/jwks-properties or some other resource. Also explain what this represents.

File: examples/authentication/server/src/auth_handler.rs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        AuthenticationHandler {
+            client_id: client_id.to_owned(),
+            client_secret: client_secret.to_owned(),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl Authentication for AuthenticationHandler {
+    /// Gets the authentication parameters needed by the client to send authorisation token
```
Comment (rbehjati) [inline examples/authentication/server/src/auth_handler.rs:57]:
```suggestion
    /// Gets the authentication parameters needed by the client to send authorisation code
```
Is an `authorisation token` the same thing as an `authorisation code`?

File: examples/authentication/server/src/token_exchanger.rs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    let claims = decode_identity_token(&result.id_token).await?;
+    info!(
+        "Validated identity token with email '{}' and subject '{}'",
+        &claims.email, &claims.sub
+    );
+    Ok(result.id_token)
+}
+
+// Decodes and validates the identity token.
+pub async fn decode_identity_token(token: &str) -> Result<Claims, Box<dyn error::Error>> {
```
Comment (rbehjati) [inline examples/authentication/server/src/token_exchanger.rs:94]: Does this function have to be public?

File: examples/authentication/server/src/token_exchanger.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        .append_pair("client_id", client_id)
+        .append_pair("client_secret", client_secret)
+        // TODO(#922): Pass the actual redirect URI from the client app.
+        .append_pair("redirect_uri", "http://127.0.0.1:8089")
+        .append_pair("grant_type", "authorization_code")
+        .finish();
+
+    // TODO(#923): Get the token URL from the discovery document.
+    let token_uri = "https://oauth2.googleapis.com/token".parse()?;
+    let response = http_client::post(token_uri, &body, "application/x-www-form-urlencoded").await?;
```
Comment (rbehjati) [inline examples/authentication/server/src/token_exchanger.rs:82]: Perhaps define a const for`"application/x-www-form-urlencoded"` somewhere?

File: examples/authentication/server/src/token_exchanger.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    let result: TokenResponse = serde_json::from_str(&response)?;
+    // Validate token.
+    let claims = decode_identity_token(&result.id_token).await?;
+    info!(
+        "Validated identity token with email '{}' and subject '{}'",
```
Comment (rbehjati) [inline examples/authentication/server/src/token_exchanger.rs:87]: I am not sure if it is a good idea to include the email address in the log.

File: examples/authentication/server/src/token_exchanger.rs (hunk 8 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Done.

File: examples/authentication/server/src/token_exchanger.rs (hunk 9 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs:91]: Done.

File: examples/authentication/server/src/token_exchanger.rs (hunk 12 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Code replaced by using `reqwest`

File: examples/authentication/server/src/token_exchanger.rs (hunk 13 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Removed. Was intended for debugging.

File: examples/authentication/server/src/token_exchanger.rs (hunk 11 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Not at the moment.

File: examples/authentication/server/src/token_exchanger.rs (hunk 7 above)
Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Removed `Serialize`. `Deserialize` is required by `decode`.

Comment (conradgrobler) [reply examples/authentication/server/src/token_exchanger.rs]: Done.

File: examples/authentication/README.md [hunk 14]
```diff
@@ -15,9 +15,14 @@ by the authentication server.
 The server can be run using:
 
 ```bash
-cargo run --package=auth_server -- --client-id=691249393555-0h52jim9ni9clkpd5chi82q9ccn44ebm.apps.googleusercontent.com
+cargo run --package=auth_server -- \
+  --client-id=691249393555-0h52jim9ni9clkpd5chi82q9ccn44ebm.apps.googleusercontent.com \
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/authentication/README.md:19]: Still what's missing are some step-by-step instructions on how to run this. Like "go to URL and copy paste this value into a file" or whatever. In that sense personally I think it's easier to leave the client ID in this command line, but up to you. Could you make sure something like this is there before submitting @conradgrobler ?

[the remark this answers is not in the document]
Comment (conradgrobler) [reply examples/authentication/README.md:19]: The values are not copied into a file, just as parameters into the command line. I reworded the text a bit and moved the content specifying the URL where the values can be copied from higher up in the page.  Hopefully this is clear enough.

**Pull request merged**